### ISSUE TITLE
Update ARIA mapping to remove instances of "?" for HTML attributes

### ISSUE THREAD

**Issue by scottaohara** [w3c/html-aam#259]
Related to issue #97, there are still instances of "?" in the document related to ARIA mappings.  

specifically:

- [ ] `contenteditable` (HTML elements) 
- [x] `span` (col; colgroup)
- [x]  `title` (HTML elements)

**Comment by joanmarie** [comment]
Regarding `contenteditable`: Should it trump the generic role? See https://github.com/w3c/aria/discussions/1449.

**Comment by scottaohara** [comment]
yes.  well.  for your specific example, yes.

for more complex `<button contenteditable>...</button>`, i honestly am not sure what should happen there.

but for generic divs, those (at least last time i checked) get exposed as textboxes.

**Comment by joanmarie** [comment]
At least in the case of Chromium, `data:text/html,<div contenteditable>foo</div>` has a computed role of `generic` when I look in the dev tools.

As a result, my (local; uncommitted) implementation in Chromium to do the ARIA 1.2 name prohibiting makes `data:text/html,<div contenteditable aria-label="name">foo</div>` not have a name. Ditto for a focusable, but not contenteditable, `div`. While that's not the best authoring, prohibiting exposure of the `aria-label` value on these interactive things also seems bad. Furthermore, I can see us getting pushback from the browser vendors.

I think one way to solve this problem is for your HTML-AAM to cause `contenteditable` (and `tabindex` >= 0) on generic elements map to a non-generic role. I'm open to other ideas. 

Note that as James verified in https://github.com/w3c/aria/discussions/1449:

> https://w3c.github.io/aria/#document-handling_author-errors_states-properties states "User Agents MUST NOT expose an ARIA attribute to the platform accessibilty API when the element's computed role prohibits use of the attribute."

So for ARIA 1.2 to exit CR, I believe we're going to need a solution ... somewhere.

**Comment by scottaohara** [comment]
Took another gander at [my test page](https://scottaohara.github.io/tests/attr-html/contenteditable.html), and yeh, don't recall if things changed, or if i was misremembering this to be more consistent - but only webkit treats `contenteditable` on a `div` as if it were a `textbox`, exposing the `AXTextArea` at the platform level.  But on other elements that expose roles (though not all) the base role is retained.  e.g., `button` and `main` continue to be exposed.  but the`aside`'s `complementary` role is dropped.

FWIW, I like your suggestion of revising the mapping for generic elements with `contenteditable` (and `tabindex=0+`).  We could take a page from Webkit and expose these elements as generic `role=group` (and indicate that an 'editable' property be consistently exposed/announced).  Then AT could potentially decide to not announce the `group` role if it remains unnamed, as they do with static unnamed `role=group`

**Comment by joanmarie** [comment]
In the review comments of my [proposed implementation](https://chromium-review.googlesource.com/c/chromium/src/+/2840267), @aleventhal said:

> Shouldn't a contenteditable get a default role of textbox. It's also generally aria-multiline="true" by default ... perhaps the user agent can compute aria-multiline via CSS properties though.

I agree that if it's a generic role and it's contenteditable, that mapping it as a textbox makes more sense than a group. Thoughts?

**Comment by joanmarie** [comment]
@jcsteh, @feerrenrut, @ggordon-vispero your input on this would be helpful. The executive summary is:

* ARIA 1.2 prohibits naming of certain roles including the new "generic" role, which is what `div`, `span`, and several other things are considered.
* That prohibition is problematic when an author sets `contenteditable` or `tabindex` on a generic (screen reader should say something when it gets focus)
* The proposed solution is that a focusable generic would be mapped as if it had the ARIA "group" role -- or possibly the "textbox" role in the case of `contenteditable`.

Do you anticipate any problems in terms of implementing this in your product?

**Comment by joanmarie** [comment]
@cookiecrook: Could you please verify that for VoiceOver an `AXGroup` is an `AXGroup` regardless of whether the element is a `div` with a `tabindex` or a `div` with `role="group"`?

**Comment by scottaohara** [comment]
@joanmarie @aleventhal understand the concern / question about mapping to `role=textbox` instead.  I do think that would make sense but for the fact that _anything_ can be a child of a `contenteditable` element, let alone set to `contenteditable`.

Presently HTML doesn't let any sort of content/elements within native text fields, nor would we expect authors to put any sort of element or role within a `role=textbox`... so to me it seemed to make far more sense to expose contenteditable elements as an "editable" `group`, rather than buck expectations and do something that isn't valid with native HTML, nor even a good idea with ARIA.

[I put together a quick test](https://codepen.io/scottohara/full/qBRQgPp) of a `role=textbox` containing text, a button, link and an `input type=text` vs the same content within a `div contenteditable` without a textbox role, the latter works better (though still not perfect, tab key consistently skips over the link within the contenteditable div - but at least what does receive focus is announced properly) right now.  With JAWS or NVDA, tabbing through the items in the `role=textbox`, none of the focusable content is announced.

**Comment by joanmarie** [comment]
@scottaohara I just tried your test case using Orca with Firefox Nightly and with Chrome stable and unstable. I can confirm the not-speaking issue with Chrome. Orca speaks all focus changes in Firefox Nightly.

Looking at Orca's debugging output, the problem is that Chrome isn't emitting the focus-change events. If Orca doesn't know focus changed, it certainly cannot announce it. I'm guessing the problem is similar/the same with JAWS and NVDA. What browser(s) did you test JAWS and NVDA with?

Regardless, independent of the what the mapping winds up being to solve the name-prohibited stuff, we need to get this Chrome bug fixed. I'll look into it later today and hopefully land a fix soon.

Once that bug is fixed so that the focus changes are announced by screen readers, are you still of the opinion that mapping a `contenteditable` `div` as a `textbox` is still a bad idea?

**Comment by scottaohara** [comment]
I tested with chrome, edge and firefox with nvda/jaws. Chromium browsers were the ones where focusing elements with the tab key were not announced.

Things seemed to work more as expected with VoiceOver - though that still wasn't perfect.

All browsers have weird quirks with focusable elements within contenteditable containers... but that's all unfortunate current behaviors that aren't contingent on AT.

regarding:
>Once that bug is fixed so that the focus changes are announced by screen readers, are you still of the opinion that mapping a contenteditable div as a textbox is still a bad idea?

I'm uncomfortable with the idea, it breaks expectations for `textbox` child allowances for both users and authors.  Every element within a `contenteditable` ancestor becomes multi-line editable, and again, any element can be set to contenteditable... so really every element should be allowed to be exposed as "editable".  So `<div role=group contenteditable>...</div>` is still going to need to be exposed as editable and I would not expect the role to be changed to `textbox`.

note: updated my previous example to add in some more content/testcases

**Comment by joanmarie** [comment]
Just filed https://bugs.chromium.org/p/chromium/issues/detail?id=1201651 and assigned it to myself.

Related asides:
* ARIA spec doesn't say `textbox` is one of the children-presentational roles.
* Also, while admittedly I am not an author, I can imagine reaching the conclusion that if I wanted to create a web equivalent of something like a [GtkSearchEntry](https://developer.gnome.org/gtk3/stable/GtkSearchEntry.html), I'd make the symbolic icon a child of the entry (which is actually what the accessibility tree looks like for that particular widget). Might not be good authoring in ARIA, but I'm not fully convinced it breaks expectations, especially given the first point.

All of that said, this chromium bug needs fixing, so I'll focus on that and wait for others to chime in on what the role of a contenteditable div should be. (And then I'll look at your additional examples. Thanks for the test cases btw!!)

**Comment by scottaohara** [comment]
@joanmarie, yes my mention of expectations was not regarding whether ARIA says child elements are presentational.  Rather, that one cannot have non-text content within native text fields ([text inputs have a content model of 'nothing' and expect no child elements](https://html.spec.whatwg.org/multipage/dom.html#concept-content-nothing) and [textarea](https://html.spec.whatwg.org/multipage/form-elements.html#the-textarea-element) only allows for text child content) in HTML.  

I'm actually curious as to why ARIA didn't specify `textbox` to have child elements as presentational... as it's relevant right now.  Consider the following (also added to my codepen).  As this is currently allowed, then what's the expectation for the button?  If `contenteditable` maps to `textbox` then I'd say that the button must be editable (can be deleted, or have its contents updated) by the user.  But, that's not logical to what the intent here likely is.
```html
<div role=textbox tabindex=0>
  Some text
  <button aria-label="clear">X</button>
</div>
```

**Comment by joanmarie** [comment]
BTW, I do get focus events (aside from links which are left out of the tab order) when the `div` with role `textbox` also has `contenteditable` set. It's only when it's missing that the focus events go missing.

And what role should a rich-text multi-line editable field (like a message composition window) get?

**Comment by scottaohara** [comment]
Great question! I think we are (hopefully) figuring that out now :)

**Comment by ggordon-vispero** [comment]
This is probably fine for JAWS assuming that the browser does the proper mapping.

It would be good to have really strongly worded language in the authoring guidelines to say to not use generic roles for things which in any way could be thought to carry semantic meaning. i.e. If you’re making an element tabbable or editable, there’s likely more semantic meaning than you think, and you should reconsider assigning a more specific role.

**Comment by carmacleod** [comment]
> what role should a rich-text multi-line editable field (like a message composition window) get?

I realize that this is not necessarily a helpful answer (or a popular one 😄 ), however in our old code editor widget, which was a contenteditable div with lots of children, I wanted to have _all the keys_, even tab (because we had some fancy indent/outdent feature), and so I had no choice but to give it `role="application"`. Of course, that was almost a decade ago? And the code editor is not really being used any more, but it worked at the time.

**Comment by cookiecrook** [comment]
> @cookiecrook: Could you please verify that for VoiceOver an `AXGroup` is an `AXGroup` regardless of whether the element is a `div` with a `tabindex` or a `div` with `role="group"`?

No, role=group has a specific AXSubrole... AXApplicationGroup if I recall correctly.

Also, I recall (but it would require more investigation than I have time for at the moment) WebKit having some heuristics for not exposing Div groups as interaction groups in some contexts. For example, it used to be that VoiceOver would force interaction (VO+Shift+Down) with any native AXGroup that had an AXTitle... So the web context `<div title="foo">` used to result in an unexpected interaction group in web interfaces... I believe we've since resolved any of that unexpected behavior, but there were many WebKit and VO changes associated with it. I don't recall them all offhand.

**Comment by aleventhal** [comment]
IMO we should use role=textbox for the contenteditable case and in parallel fix any issues with ```<div contenteditable role="textbox">```. It's a common use case, and I don't know of anything in the spec to  state that it's wrong or that role="textbox" means something is plain text inside. 
For example, Google Docs uses: ```<div tabindex="0" spellcheck="false" role="textbox" aria-multiline="true" aria-label="Document content" contenteditable="true">``` 
In this case, you can easily have markup for links inside, and although the links are not tabbable while in an editable view (as opposed to published view), they are editable and actionable.

Scott, thanks for finding the issues with that.

**Comment by scottaohara** [comment]
Thanks @aleventhal, if mapping to `textbox` is not causing any issues there, then that's definitely a positive.  I suppose another "pro" for mapping to `textbox` would mean that screen readers would be more apt to go into forms mode automatically, than if it were mapped to `group` which also happened to be editable....

But I'm now wondering if `contenteditable` should impact more roles than just changing a `generic` to a `textbox`.  For instance, why not overwrite other non-interactive roles and turn those into `textbox` elements as well?  Does that need to be considered?

```html
<div role=group | region | banner | main | paragraph | whatever contenteditable>
  all the buttons and inputs and whatever
</div>
```

The above, regardless of the chosen role, still needs to be exposed as editable.  Should the role always be remapped to a `textbox` then?  Otherwise there's still a need to make sure an explicit`role=group` with `contenteditable` is announced as editable anyway. Paragraphs are naming prohibited, but if they're `contenteditable` then they, and other elements that prohibit naming, would also need to be implicit `role=textbox`.  That would happen even if we mapped to `group` too, but it does make certain markup patterns that you wouldn't think to nest textboxes within make more sense.

While `role=textbox` might be fine for google docs, i think it's safe to say not all the different types of elements that could be found within a web page can exist in a google doc.  So I think it makes sense to look at unsafe situations of use as well.

As a quick example - the following markup is very much allowed.  While mapping to `group` or `textbox` wont' make much of a difference for the `button` since it treats child elements as presentational.  

```html
<button>
  <span contenteditable>My button label</span>
</button>

<a href="#">
  <span contenteditable>My link label</span>
</a>
```

However, mapping to `textbox` would be odd within a link. And if it was an `<input>` or `<textarea>` in either case above, that would be invalid HTML.


Anyway, I don't think I have any more arguments against `textbox` than what I've said before. If it can be mapped / will behave in a way that won't be quirky (or quirkier than letting _any_ HTML element be set to `contenteditable` already is), then let's go with `textbox`.   I think if we do that though, it stands to reason that maybe `generic` isn't the only role that gets remapped if `contenteditable` is set.

Thanks

**Comment by joanmarie** [comment]
> > @cookiecrook: Could you please verify that for VoiceOver an `AXGroup` is an `AXGroup` regardless of whether the element is a `div` with a `tabindex` or a `div` with `role="group"`?
> 
> No, role=group has a specific AXSubrole... AXApplicationGroup if I recall correctly.

Oh yeah.... Forgot about that. Thanks!

That said, using Safari Technology Preview with VoiceOver, I am not seeing any behavior difference between these:
* `<div tabindex="0">foo</div>`
* `<div tabindex="0" role="group">foo</div>`

So if the author did the first one (sans role), but it was mapped as if the author had done the second one, would this in any way break VO's expectations or the user experience?

**Comment by joanmarie** [comment]
@scottaohara my vote would be to not allow `contenteditable` to trump non-generic roles.

 `<button contenteditable>foo</button>` is still visually a button and its buttoniness is presumably relevant/necessary. If that turned into a `textbox`, the user loses meaning. (Or perhaps more accurately, the screen readers lose the ability to handle that unexpected situation and present it to the user.)

**Comment by aleventhal** [comment]
Agree with Joanie.

@scottaohara thanks again for all the deep thought into this.

**Comment by scottaohara** [comment]
@joanmarie yup.  I wasn't suggesting that.  i was questioning non-interactive roles

like `<p contenteditable>`.  or like the examples I have in my last post of spans with contenteditable inside of links or buttons (though spans would be generic too).

**Comment by joanmarie** [comment]
Hmmm. If we wanted to allow `<p contenteditable aria-label="Tell me a story"></p>` and not cause the name to be stripped out, then I suppose we would because name is prohibited on paragraphs too. Yay?

**Comment by scottaohara** [comment]
yup.  any naming prohibited element / role with contenteditable needs to have its role changed, essentially.

but per your comment "is still visually a button and its buttoniness is presumably relevant/necessary." the same to could be said about anything that contenteditable can be applied to (which is everything).

**Comment by joanmarie** [comment]
The other possibility is changing the ARIA spec to state that name prohibition doesn't apply when something is tab-focusable or contenteditable.

**Comment by joanmarie** [comment]
Interesting scenario (I'm making this change in Chromium locally and seeing what "breaks"): 

`<div contenteditable aria-readonly="true">`

When the `div` above is mapped to `generic` or `group`, `aria-readonly` isn't exposed; when it's mapped to `textbox`, it is.

I forget who trumps what in that scenario (i.e. if the native thing is really editable, should we be telling ATs otherwise?). And I don't have time to look at the moment. My main point is that there is web engine logic kicking in as a consequence of the implicit role of contenteditable elements being more closely related to the functionality of the native element.

ARIA WG meaty-topic topic? :smile:

**Comment by kbabbitt** [comment]
I think mapping contenteditable to role=textbox makes sense, but while reviewing @joanmarie's change for it in Chromium, I noticed one downstream effect on UIA.

Core AAM calls for role=textbox to be mapped to the UIA Edit control type even for multiline [1]. But the docs for Edit state that it's intended for single-line, plain-text editing [2], and the docs for the Value pattern - which Edit is required to support - state that multi-line controls do not implement Value [3].

With contenteditable being mapped to role=textbox, that means a div with no other role will show up as the Edit control type rather than Group as it does currently, which might be misleading to a screen reader.

It seems to me we should change Core AAM to map multi-line textbox to Group control type with Text control pattern, but I don't know the full history of the current mapping. @travisleithead do you have any thoughts on this?

[1] https://www.w3.org/TR/core-aam-1.1/#role-map-textbox
[2] https://docs.microsoft.com/en-us/windows/win32/winauto/uiauto-supporteditcontroltype
[3] https://docs.microsoft.com/en-us/windows/win32/api/uiautomationcore/nf-uiautomationcore-ivalueprovider-get_value

**Comment by jcsteh** [comment]
I agree with @joanmarie: I think generics with contenteditable should map to role="textbox", but non-generics with contenteditable should not. I realise that raises some tricky questions about things like `<p contenteditable>`. Anecdotally, though, I've seen more generics with contenteditable in the wild than non-generics.

ATs already have to support the "editable" state (aside from the textbox role) now to handle contenteditable and I don't think that should change. However, from a screen reader user perspective, `<div contenteditable>` is reasonably common, but "section editable" (what gets announced by NVDA in this case) is far less intuitive than "edit" (what gets announced by NVDA for `role="textbox"`).

All of that said, one possible issue I see with respect to browser implementation is that this might mean that the role of an element changes when contenteditable is enabled/disabled. We already have cases of that right now, but introducing more of those feels bad, especially given that changing the role might end up rebuilding the subtree (I can't remember the latest spec guidance for role changes) and the subtree for an editor could be massive.

**Comment by scottaohara** [comment]
Adding in @joanmarie's [issue for the ARIA spec](https://github.com/w3c/aria/issues/1493) concerning adding more normative guidance for the `textbox` role.   Overlaps with this conversation.

Playing around with https://codepen.io/scottohara/pen/KKoWbzo lead to some interesting results.  

Finding that some elements have their roles suppressed and announce as "edit".  Others, like `<div role=button>` or `<h3>` do not consistently have their roles suppressed, and do not consistently announce as editable.  

I'm also wondering how beneficial it is to allow some focusable elements (but not all - see above comments about links not being focusable within a contenteditable container) continue to be focusable.  Navigating to a div contenteditable with a nested button element, it may not be clear to someone that they're still within the editable area - especially since in all other non-contenteditable cases using native HTML elements, you will not find buttons within text fields.  

What say people to negating the focusability of elements within a contenteditable container?

What is the stance on what elements become role=textbox vs just being announced as "edit".  e.g. if generic becomes textbox, does article as well?  do headings remain headings announced as editable?  or do they too become textboxes?

**Comment by ggordon-vispero** [comment]
I think that this will negatively impact tools like Quip and maybe even Microsoft Loop where contentEditables sometimes serve two purposes. You can edit, but you can also take action  using the contained controls.

**Comment by scottaohara** [comment]
spoke with @ggordon-vispero about the topic of what to map `<div contenteditable>` to recently.

We both had similar concerns that mapping to `role=textbox` would be inconsistent with how the attribute is treated on other elements.  For instance,

```html
<article contenteditable>
  <!-- paragraphs, lists, links and buttons here -->
</article>
```
vs
```html
<div contenteditable>
  <!-- paragraphs, lists, links and buttons here -->
</div>
```

have the potential to be handled very differently if the former retains it's article role, but the latter - even though it would contain similar types of content - is mapped to behave like `<input type=text>`.

**Comment by ggordon-vispero** [comment]
To build on Scott’s comments: if a browser maps a contenteditable div to a specific role like Editable Grouping and a screen reader doesn’t want to expose that role to users, it’s free to map  it to a  more general one like textbox. But if the browser itself maps to textbox,  a screen reader doesn’t  have enough information to go the other way.
